src/bloom.cpp

```diff
@@ -22,35 +23,36 @@
 
 using namespace std;
 
-CBloomFilter::CBloomFilter(unsigned int nElements, double nFPRate, unsigned int nTweakIn, unsigned char nFlagsIn) :
-    /**
-     * The ideal size for a bloom filter with a given number of elements and false positive rate is:
-     * - nElements * log(fp rate) / ln(2)^2
-     * We ignore filter parameters which will create a bloom filter larger than the protocol limits
-     */
-    vData(min((unsigned int)(-1  / LN2SQUARED * nElements * log(nFPRate)), MAX_BLOOM_FILTER_SIZE * 8) / 8),
-    /**
-     * The ideal number of hash functions is filter size * ln(2) / number of elements
-     * Again, we ignore filter parameters which will create a bloom filter with more hash functions than the protocol limits
-     * See https://en.wikipedia.org/wiki/Bloom_filter for an explanation of these formulas
-     */
-    isFull(false),
-    isEmpty(false),
-    nHashFuncs(min((unsigned int)(vData.size() * 8 / nElements * LN2), MAX_HASH_FUNCS)),
-    nTweak(nTweakIn),
-    nFlags(nFlagsIn)
-{
+void CBloomFilter::setup(unsigned int nElements, double nFPRate, unsigned int nTweakIn, unsigned char nFlagsIn, bool size_constrained) {
+    if (nElements == 0) {
+        LogPrintf("Construction of empty CBloomFilter attempted.\n");
+        nElements = 1;
+    }
+    unsigned desired_vdata_size = (unsigned int)(-1  / LN2SQUARED * nElements * log(nFPRate) / 8);
+
+    if (size_constrained)
+        desired_vdata_size = min(desired_vdata_size, MAX_BLOOM_FILTER_SIZE);
+    
+    vData.resize(desired_vdata_size, 0);
+    isFull = vData.size() == 0;
+    isEmpty = vData.size() > 0;
```

Comment:
Also noted that the default constructor left isEmpty at false, which was essentially the same inconsistency. Fixed that now as well.